src/lib.rs

```diff
@@ -103,237 +106,129 @@ const VERSION: &str = "0.0.1";
 const LANG: &str = "rust";
 
 #[doc(hidden)]
-pub trait ConnectionState: private::Sealed {}
-mod private {
-    pub trait Sealed {}
-    impl Sealed for super::NotConnected {}
-    impl Sealed for super::Authenticated {}
-    impl Sealed for super::Connected {}
-    impl Sealed for dyn super::OptionState {}
+pub mod options_typestate {
+    /// `ConnectionOptions` typestate indicating
+    /// that there has not yet been
+    /// any auth-related configuration
+    /// provided yet.
+    #[derive(Debug, Copy, Clone, Default)]
+    pub struct Unauthenticated;
+
+    /// `ConnectionOptions` typestate indicating
+    /// that auth-related configuration
+    /// has been provided, and may not
+    /// be provided again.
+    #[derive(Debug, Copy, Clone)]
+    pub struct Authenticated;
+
+    /// `ConnectionOptions` typestate indicating that
+    /// this `ConnectionOptions` has been used to create
+    /// a `Connection` and may not be changed.
+    #[derive(Debug, Copy, Clone)]
+    pub struct Finalized;
 }
 
-impl ConnectionState for NotConnected {}
-impl ConnectionState for Authenticated {}
-impl ConnectionState for Connected {}
-
-/// A NATS connection.
-#[derive(Debug)]
-pub struct Connection<S: ConnectionState> {
-    options: Options,
-    state: S,
+/// A configuration object for a NATS connection.
+#[derive(Clone, Debug, Default)]
+pub struct ConnectionOptions<TypeState = options_typestate::Unauthenticated> {
```

Comment:
This might look kind of weird to people, it's the specification of a default generic type. In practice it's not as helpful as it seems like it should be, but from time to time it can help reduce friction. 

One that you may have used already is std's HashMap for its 3rd generic parameter (the hashing implementation) that you can omit due to the fact that it's default.

Reply:
Not sure this is better than what we had to start TBH..

Reply:
Which aspect? I'm blocked on implementing reconnection logic because these refactors dramatically improve my ability to reason about the Connection object.

Reply:
Understood but the client seems just as complex to me now as it was before.